### ISSUE TITLE
How to trace across multiple DCs (each having its own data backend)

### ISSUE THREAD

**Issue by raffraffraff** [jaegertracing/jaeger#1122]
## Requirement - what kind of business use case are you trying to solve?
I'm planning to deploy Jaeger to a large multi-datacenter environment. I would like to create a separate Cassandra cluster in each datacenter. However, this would prevent us from being able to trace requests that get routed from one datacenter to another.

## Problem - what in Jaeger blocks you from solving the requirement?
Jaeger UI cannot search across multiple Cassandra clusters

## Proposal - what do you suggest to solve the problem or improve the existing situation?
I don't have one :(

**Comment by yurishkuro** [comment]
Does your company have an established Kafka infrastructure with cross-zone replication?  This would be one way to deal with multi-zone setup by having one global Jaeger installation (which you can duplicate in another zone if you want an HA backup).

**Comment by rwkarg** [comment]
We are in a similar situation (cross-DC services needing tracing).
We do have established experience with Kafka. 
I'm assuming that this would be best implemented once the Ingest? services (Kafka -> DB connectors) are available?

**Comment by yurishkuro** [comment]
@rwkarg Kafka ingesters are available since 1.7 (just not documented yet)

**Comment by raffraffraff** [comment]
> This would be one way to deal with multi-zone setup by having one global Jaeger installation (which you can duplicate in another zone if you want an HA backup).

A Kafka solution was my initial thought, didn't know that ingesters were a thing yet. One thing I'd like to do is keep the cross-DC chatter to a minimum. For instance, how about having a completely separate Jaeger installations, but using Kafka to send _only_ the traces that are relevant to the that (ie: traces that originated in DC1 should get sent back to DC1 and ingested into its local Cassandra cluster).

**Comment by yurishkuro** [comment]
Yes, this is the long term solution we are thinking about, happy to brainstorm it more. It is not trivial though, as there is no clear signal in the traces themselves in one zone about whether there are also spans in another zone. Broadcasting trace IDs with zone identity to all zones is a possible solution, but somewhat expensive.

**Comment by raffraffraff** [comment]
How about allowing the Jaeger UI / Query to search across multiple Cassandra clusters? That way, there's no need to do any cross-site replication (one of the things I'd love to avoid, if possible)

**Comment by yurishkuro** [comment]
Query federation is the easiest to implement, but it is also least useful. The main value of tracing is in aggregate analysis, which doesn't scale if all queries fan out.

**Comment by tylerchr** [comment]
To me, something like the suggestion from @raffraffraff sounds like it would make Jaeger most useful. It seems crazy to transmit all this trace data over the Internet, considering that the expected value of any randomly-selected trace is so near zero (why use bandwidth for traces that are so unlikely to be looked at?). It would be neat if all traces that _originated_ in a given zone were fully kept in that zone, but that's tricky for the reasons you described and also subject to my logic above.

I think by "aggregate analysis" you mean gathering together all the spans from a trace. It is true that speculatively searching all the zones for any spans they might have doesn't scale well, but it's quite convenient for smaller installations (in my case we'd have maybe a half-dozen of those). Perhaps this could be mitigated by keeping an index of which zones have spans for which trace IDs, possibly built using the "Broadcasting trace IDs" system you mentioned, such that jaeger-query only has to talk to the zones that actually have span data for the trace in question? That seems nice because then raw span data only moves crosses between DCs at query-time, when it's obvious that I'm interested in it, rather than collection time, when the probability of me later looking at any given trace is very low. One downside is that the availability of a trace via jaeger-query's API is delayed until the trace ID broadcast has occurred.

**Comment by yurishkuro** [comment]
> t seems crazy to transmit all this trace data over the Internet, considering that the expected value of any randomly-selected trace is so near zero (why use bandwidth for traces that are so unlikely to be looked at?)

So precisely because of low value of individual traces we want to run aggregations, and aggregations do look at all data eagerly. So I don't think you can avoid transmitting it between zones. What we can and do want to avoid is replicating traces that are fully local to a given zone.


This is not to say there's no value in building query-time federation, it's just much lower priority (for us, at least) and even redundant if sensible replication is implemented. But smaller setups that don't want to deal with replication and just want a single multi-zone UI could benefit from query federation.

**Comment by raffraffraff** [comment]
FWIW: In my case, most of the traffic stays within a DC, and we have a _lot_ of DCs, so we don't want to replicate all-to-all. So query-time federation would actually suit me.

**Comment by yurishkuro** [comment]
Another option we discussed elsewhere is using baggage to propagate the zones encountered during the transaction execution. It makes it explicit when looking at the trace which zones the request travelled through. But the cons are that it requires changes to clients to:

* know the zone they are running in (via configuration)
* adding it to the baggage automatically & concatenating when multiple zones are visited
* recording that value in the span (that happens automatically today when baggage is set)

**Comment by tiffon** [comment]
@yurishkuro How would early spans learn of zones from later spans?

**Comment by yurishkuro** [comment]
When the full trace is available, e.g. in the "trace_completed" Flink job, we will know that there are spans in other zones.  Technically this does not even require full-stack baggage, just a single-hop header that informs the receiver that the call came from a different zone. We could also use the response headers to pass the zone information back to the caller. But neither of these still cover the whole call graph, so only in post-processing can we know that there spans in other regions.

The benefit of doing it in post-processing is that it's on-demand, which was one of the concerned raised earlier about eager replication. If we instead want to proactively replicate spans to all zones encountered in the trace, then the broadcasting solution is much easier to implement. 

I was thinking of how to address the race conditions in the broadcast approach, e.g. when a zone X receives a broadcast of trace id T saying that zone Y is interested in that trace, then rather than only replicating spans received from that point on, zone X needs to go back and replicate spans that may have already been stored.

**Comment by jerishsd** [comment]
+1 (Have the exact same requirement :))

**Comment by hobbytp** [comment]
+1 :-)

**Comment by sankareds** [comment]
+1

**Comment by aravindvs** [comment]
Any updates on this? Query federation will be super useful to avoid cross-region traffic.
